### PR TITLE
Adding ensemble

### PR DESCRIPTION
**Pull request by xisen-w**
## Description
Added ensemble component following the pattern of data loader component. This includes:
- New ensemble component with CoSTEER implementation
- Test files and evaluation logic
- Integration with aerial-cactus-identification competition workflow
- Specification handling for ensemble tasks

## Motivation and Context
The ensemble component is needed to combine multiple model predictions in a systematic way, following the same pattern as other components (data loader, feature engineering, model). This ensures consistency in the codebase while adding new functionality for model ensemble decisions.

## How Has This Been Tested?
- [x] If you are adding a new feature, test on your own test scripts.
Screenshot 2024-12-13 at 1 34 56 PM
Screenshot 2024-12-13 at 1 34 44 PM

The component has been tested with:
- Unit tests in eval_tests/ensemble_test.py
- Integration testing with the aerial-cactus-identification competition workflow
- Test cases covering both synthetic competition data

### CODE REVIEW

File: rdagent/components/coder/data_science/ensemble/eval.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+from rdagent.core.evolving_framework import QueriedKnowledge
+from rdagent.core.experiment import FBWorkspace, Task
+from rdagent.oai.llm_utils import APIBackend
+from rdagent.utils.agent.tpl import T
+from rdagent.utils.env import DSDockerConf, DockerEnv
+
+DIRNAME = Path(__file__).absolute().resolve().parent
+
+@dataclass
+class EnsembleEvalFeedback(Feedback):
```
Comment (you-n-g) [inline rdagent/components/coder/data_science/ensemble/eval.py:20]: Use the more general version provided by bowen

File: rdagent/components/coder/data_science/ensemble/prompts.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    {{ ensemble_spec }}
+
+    Please respond with the code in the following json format:
+    {
+        "code": "The Python code as a string."
+    }
+  user: |-
+    Please implement an ensemble function with the following specification:
+    ```python
+    def ens_and_decision(test_pred_l: list[np.ndarray], val_pred_l: list[np.ndarray], val_label: np.ndarray) -> np.ndarray:
```
Comment (you-n-g) [inline rdagent/components/coder/data_science/ensemble/prompts.yaml:19]: You should not hard code this here.   Please read it from the spec dynamically

File: rdagent/components/coder/data_science/ensemble/eval.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        queried_knowledge: QueriedKnowledge = None,
+        **kwargs,
+    ) -> CoSTEERSingleFeedback:
+        
+        target_task_information = target_task.get_task_information()
+        if (queried_knowledge is not None and
+                target_task_information in queried_knowledge.success_task_to_knowledge_dict):
+            return queried_knowledge.success_task_to_knowledge_dict[target_task_information].feedback
+        elif queried_knowledge is not None and target_task_information in queried_knowledge.failed_task_info_set:
+            return CoSTEERSingleFeedback(
```
Comment (you-n-g) [inline rdagent/components/coder/data_science/ensemble/eval.py:42]: Please update the code to make sure you return the right version of class

File: rdagent/components/coder/data_science/ensemble/eval.py (hunk 1 above)
Comment (xisen-w) [reply rdagent/components/coder/data_science/ensemble/eval.py]: This is now resolved.

File: rdagent/components/coder/data_science/ensemble/prompts.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    {{ competition_info }}
+
+    Please respond with the code in the following json format:
+    {
+        "code": "The Python code as a string."
+    }
+  user: |-
+    Please implement an ensemble function with the following specification:
+    -----------Ensemble Specification-----------
+    {{ ensemble_spec }}
```
Comment (WinstonLiyt) [inline rdagent/components/coder/data_science/ensemble/prompts.yaml:16]: image
Like this

**Pull request merged**